contributors/design-proposals/statefulset-update.md

```diff
@@ -0,0 +1,847 @@
+    `.Status.UpdatedTemplateGeneration`.
+    1. All Pods in the StatefulSet have been generated from the PodTemplate 
+    labeled with a `StatefulSetTemplateGenerationLabel` equal to its 
+    `.Status.TemplateGeneration`.
+1. If the StatefulSet's `.Spec.GenerationPartition` is not nil, then the 
+following is true.
+    1. All Pods with ordinals is the sequence `[0,.Spec.GenerationPartition)` have 
+     been generated from the PodTemplate in the StatefulSet's revision history
+     that is labeled with a `StatefulSetTemplateGenerationLabel` equal to 
+     `.Status.TemplateGeneration`.
+    1. All Pods with ordinals in the sequence 
+     `[Spec.GenerationParition,.Spec.Replicas)` have been created with the 
+     PodTemplate in the StatefulSet's revision history that is labeled with a 
+     `StatefulSetTemplateGenerationLabel` equal to 
+     `.Status.UpdatedTemplateGeneration`.
+
+### Revised Controller Algorithm
+The StatefulSet controller will use the following algorithm to continue to 
+make progress toward the user declared [target state](#target-state) while 
+respecting the controller's 
+[identity](https://kubernetes.io/docs/concepts/abstractions/controllers/statefulsets/#pod-identity), 
+[deployment, and scaling](https://kubernetes.io/docs/concepts/abstractions/controllers/statefulsets/#deployment-and-scaling-guarantee) 
+guarantees.
+
+1. The controller will 
+[reconstruct the revision history](#history-reconstruction) of the StatefulSet.
+1. The controller will process any [template updates](#template-updates) to 
+ensure that the StatefulSet's revision history is consistent with the user 
+declared desired state.
+1. The controller will select all Pods in the StatefulSet, filter any Pods not 
+owned by the StatefulSet, and sort the remaining Pods in ordinal order.
+1. If any Pods with ordinals in the sequence `[0,.Spec.Replicas)` have not been 
+created, for the Pod corresponding to the lowest such ordinal, the controller 
+will [select the PodTemplate](#podtemplate-selection) from the StatefulSet's 
+revision history corresponding to the Pod's ordinal and create the Pod.
+1. If all Pods in the sequence `[0,.Spec.Replicas)` have been created, but if any 
+have a do not have a Ready Condition, the StatefulSet controller will wait for 
+these Pods to either become Ready, or to be completely deleted. 
+1. If all Pods in the sequence `[0,.Spec.Replicas)` have a Condition indicating 
+Ready, and if `.Spec.Replicas` is less than `.Status.Replicas`, the controller 
+will delete the Pod corresponding to the largest ordinal. This implies that 
+scaling takes precedence over Pod updates.
+1. If all Pods in the range `[0,.Spec.Replicas)` have a Status of Running and 
+a Ready Condition, if `.Spec.Replicas` is equal to `.Status.Replicas`, and if 
+there are Pods that do not match the 
+[declared desired PodTemplate](#podtemplate-selection), the Pod corresponding to 
+the largest ordinal will be deleted.
```

Comment:
There must be a more subtle reason than just, "We can't terminate N while >N are Ready" because you also violate that invariant anyway while the rollout is in progress (e.g. update Pod-4 after Pod-5 has been updated and is Ready again).

Can you elaborate in the doc on why you propose to start at the high end and work down? Maybe an example in terms of a specific app that you thought about as a use case.

Also, in earlier offline discussions, we talked about giving the user the option of updating 0->N or N->0, but I don't see mention of that here. Is that just because it won't be part of the initial proposal (which is fine), or because you thought of a reason it will never make sense?